pyquil/reference_simulator.py

```diff
@@ -163,8 +186,40 @@ class ReferenceDensitySimulator(AbstractQuantumSimulator):
     def __init__(self, n_qubits: int, rs: RandomState = None):
         self.n_qubits = n_qubits
         self.rs = rs
-        self.density = np.zeros((2 ** n_qubits, 2 ** n_qubits), dtype=np.complex128)
-        self.density[0, 0] = complex(1.0, 0)
+        self.set_initial_state(None).reset()
+
+    def set_initial_state(self, state_matrix):
+        """
+        This method is the correct way (TM) to update the initial state matrix that is
+        initialized every time reset() is called. The default initial state of
+        ReferenceDensitySimulator is |000...00>.
+
+        Note that the current state matrix, i.e. ``self.density`` is not affected by this
+        method; you must change it directly or else call reset() after calling this method.
+
+        To restore default state initialization behavior of ReferenceDensitySimulator pass in
+        ``state_matrix = None`` and then call reset().
+
+        :param state_matrix: numpy.ndarray or None.
+        :return: ``self`` to support method chaining.
+        """
+        if state_matrix is None:
```

Comment:
I am a little skeeved out by the use of `None` to indicate a particular special construction. I'd prefer that there were a helper function `manufacture_ground_state` and that the user was expected to pipe its output through to `set_initial_state`.

Reply:
- [x] to do.

Good point. 

I made the helper function you suggested but gave it a different name. I . I was toying with `n_qubit_zero_state` but ended up going with 'zero_state_matrix` as the latter name indicates it is a matrix (not a vector) and the number of qubits is the only argument to the function so the code is human readable e.g.

```python
n_qubits = 4
rho = zero_state_matrix(n_qubits)
```

Let me know what you think.